src/main/java/com/yelp/nrtsearch/server/luceneserver/IndexState.java

```diff
@@ -731,6 +727,60 @@ public synchronized long commit() throws IOException {
     JsonObject saveState = new JsonObject();
     saveState.add("state", getSaveState());
     saveLoadState.save(saveState);
+    SnapshotId snapshotId = null;
+    try {
+      if (this.getShard(0).isPrimary()
+          && globalState.getIncArchiver().isPresent()
+          && commitRequest.isPresent()) {
+        CreateSnapshotRequest createSnapshotRequest =
+            CreateSnapshotRequest.newBuilder()
+                .setIndexName(commitRequest.get().getIndexName())
+                .build();
+
+        snapshotId =
+            new CreateSnapshotHandler().createSnapshot(this, createSnapshotRequest).getSnapshotId();
+        // upload data
+        Collection<String> segmentFiles = getSegmentFilesInSnapshot(this, snapshotId);
+        String resourceData = IndexBackupUtils.getResourceData(this.name);
+        String versionHash =
+            globalState
+                .getIncArchiver()
+                .get()
+                .upload(
+                    globalState.configuration.getServiceName(),
+                    resourceData,
+                    this.rootDir,
+                    segmentFiles,
+                    Collections.emptyList(),
+                    true);
+        globalState
+            .getIncArchiver()
+            .get()
+            .blessVersion(globalState.configuration.getServiceName(), resourceData, versionHash);
+        // upload metadata
+        String resourceMetadata = IndexBackupUtils.getResourceMetadata(this.name);
+        versionHash =
+            globalState
+                .getIncArchiver()
+                .get()
+                .upload(
+                    globalState.configuration.getServiceName(),
+                    resourceMetadata,
+                    this.globalState.stateDir,
+                    Collections.emptyList(),
+                    Collections.emptyList(),
+                    true);
+        globalState
+            .getIncArchiver()
+            .get()
+            .blessVersion(
+                globalState.configuration.getServiceName(), resourceMetadata, versionHash);
```

Comment:
The way I saw this was that the contract with the client would be a `commit` ensures that all state and data thus far is "committed".  Calling Backup will also do the same as of now. We  could make it explicit that
- `commit` only commits data and 
- `backup` call is needed to commit both data and state. 

 Although, I am not entirely sure what benefits we get by not having the state committed to s3 on commit as well and having this additional requirement of using backup on the client.

Reply:
Commit is called much more frequently than backup, so state would be written to S3 much more frequently and even if it hasn't changed. I think it would just be more efficient to not do that.
The updated contract looks good to me.

Reply:
> The updated contract looks good to me.
 I am not sure I understand what you mean? Are you saying that you are okay with using `Commit` to commit both data and state and deprecate `Backup` as I suggested?

Reply:
I meant we can do this:
- `commit` only commits data and
- `backup` call is needed to commit both data and state.

Reply:
Ok, this one I strongly have an opinion on :)

I would rather have only one `commit` endpoint that does the backup for us.  Like I mentioned before it is then transparent to client and they do not have to worry about differences  of nuances between backup of state/data and use two endpoints. Previously we used backup since we had no way to commit to s3 that frequently. Now with incremental backups we can backup all required data and metadata and not let client be burdened with it.

Regarding the point you mentioned about 
`so state would be written to S3 much more frequently and even if it hasn't changed. I think it would just be more efficient to not do that.`

State files are merely a few hundred bytes so I am not sure why it is not efficient to upload them often even though they have not changed. Simpler client contract (a single way to backup) trumps uploading a few files redundantly imo. Also if we are worried about files being on s3 for a long time we can easily have the older ones removed by running cron jobs to purge them.

Reply:
As discussed in person we will move ahead with using `commit` as the single command to backup all data and state and deprecate usage of Backup.